Content.Server/Power/SMES/PowerSmesSystem.cs

```diff
@@ -1,16 +1,75 @@
+using Content.Server.Power.Components;
+using Content.Server.Power.EntitySystems;
+using Content.Shared.Power;
+using Content.Shared.Rounding;
+using Content.Shared.SMES;
 using JetBrains.Annotations;
+using Robust.Shared.Timing;
 
 namespace Content.Server.Power.SMES
 {
     [UsedImplicitly]
     internal sealed class PowerSmesSystem : EntitySystem
     {
-        public override void Update(float frameTime)
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+        [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            UpdatesAfter.Add(typeof(PowerNetSystem));
+
+            SubscribeLocalEvent<SmesComponent, MapInitEvent>(OnMapInit);
+            SubscribeLocalEvent<SmesComponent, ChargeChangedEvent>(OnBatteryChargeChanged);
+        }
+
+        private void OnMapInit(EntityUid uid, SmesComponent component, MapInitEvent args)
+        {
+            UpdateSmesState(uid, component);
+        }
+
+        private void OnBatteryChargeChanged(EntityUid uid, SmesComponent component, ChargeChangedEvent args)
+        {
+            UpdateSmesState(uid, component);
+        }
+
+        private void UpdateSmesState(EntityUid uid, SmesComponent smes)
         {
-            foreach (var comp in EntityManager.EntityQuery<SmesComponent>(true))
+            var newLevel = CalcChargeLevel(uid);
+            if (newLevel != smes.LastChargeLevel && smes.LastChargeLevelTime + SmesComponent.VisualsChangeDelay < _gameTiming.CurTime)
             {
-                comp.OnUpdate();
+                smes.LastChargeLevel = newLevel;
+                smes.LastChargeLevelTime = _gameTiming.CurTime;
+
+                _appearance.SetData(uid, SmesVisuals.LastChargeLevel, newLevel);
+            }
+
+            var newChargeState = CalcChargeState(uid);
+            if (newChargeState != smes.LastChargeState && smes.LastChargeStateTime + SmesComponent.VisualsChangeDelay < _gameTiming.CurTime)
+            {
+                smes.LastChargeState = newChargeState;
+                smes.LastChargeStateTime = _gameTiming.CurTime;
+
+                _appearance.SetData(uid, SmesVisuals.LastChargeState, newChargeState);
             }
         }
+
+        private int CalcChargeLevel(EntityUid uid)
+        {
+            var battery = Comp<BatteryComponent>(uid);
+            return ContentHelpers.RoundToLevels(battery.CurrentCharge, battery.MaxCharge, 6);
+        }
+
+        private ChargeState CalcChargeState(EntityUid uid)
+        {
+            var netBattery = Comp<PowerNetworkBatteryComponent>(uid);
+            return (netBattery.CurrentSupply - netBattery.CurrentReceiving) switch
+            {
+                > 0 => ChargeState.Discharging,
+                < 0 => ChargeState.Charging,
+                _ => ChargeState.Still
+            };
```

Comment:
use resolve pattern here, i.e
```suggestion
        private int CalcChargeLevel(EntityUid uid, BatteryComponent? component = null)
        {
            if (!Resolve<BatteryComponent>(uid, ref component)
            	return 0;
            	
            return ContentHelpers.RoundToLevels(battery.CurrentCharge, battery.MaxCharge, 6);
        }
```

Reply:
Thanks - I found it a bit confusing to see the pattern of optional args which are never filled, but I guess its used in this codebase as a shorthand for defining variables?

Reply:
Done.